packages/web/src/components/premium-content-purchase-modal/PremiumContentPurchaseModal.tsx

```diff
@@ -1,38 +1,68 @@
+import { useCallback, useEffect, useState } from 'react'
+
 import {
   useGetTrackById,
   usePremiumContentPurchaseModal,
-  useUSDCBalance
+  useUSDCBalance,
+  buyUSDCActions,
+  buyUSDCSelectors
 } from '@audius/common'
 import { IconCart, Modal, ModalContent, ModalHeader } from '@audius/stems'
+import { useDispatch, useSelector } from 'react-redux'
 
 import { Icon } from 'components/Icon'
 import { Text } from 'components/typography'
 
 import styles from './PremiumContentPurchaseModal.module.css'
 import { PurchaseContentForm } from './components/PurchaseContentForm'
 
+const { startRecoveryIfNecessary, recoveryStatusChanged } = buyUSDCActions
+const { getRecoveryStatus } = buyUSDCSelectors
+
 const messages = {
   completePurchase: 'Complete Purchase'
 }
 
 export const PremiumContentPurchaseModal = () => {
+  const dispatch = useDispatch()
   const {
     isOpen,
     onClose,
     onClosed,
     data: { contentId: trackId }
   } = usePremiumContentPurchaseModal()
 
-  const { data: balance } = useUSDCBalance()
+  const { data: balance, refresh } = useUSDCBalance()
   const { data: track } = useGetTrackById(
     { id: trackId! },
     { disabled: !trackId }
   )
 
+  const [shouldAttemptRecovery, setShouldAttemptRecovery] = useState(true)
+  const recoveryStatus = useSelector(getRecoveryStatus)
+
+  useEffect(() => {
+    if (trackId && shouldAttemptRecovery) {
```

Comment:
What's the goal here? Attempt recovery when the modal loads? Attempt it each time the track changes?
I'm wondering if we need the boolean at all.

Reply:
Got it. So if you remove the trackId and shouldAttemptRecovery dependencies, it's going to run once on mount of the modal and then never again.

Reply:
Or you could just do the trackId dependency and check that it's a valid trackId. As that will only change when the modal is re-initialized for a new track.
Alternatively, if you push this down into the <Form /> component and do what I said above with no dependencies, then you will get the recovery only when the form mounts.

(I may have led you astray on the first suggestion since I believe our modals are rendered in a way where they are mounted at app startup and never unmounted)